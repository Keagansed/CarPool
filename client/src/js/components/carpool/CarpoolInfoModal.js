// File Type: Component

import { Link } from 'react-router-dom';
import React, { Component } from 'react';

import app from '../../stores/MessagingStore'
import { getFromStorage } from '../../utils/localStorage.js'

const display = {
    display: 'block'
};
const hide = {
    display: 'none'
};

/*
 * Purpose: an interface that displays the members of the carpool and a link to their profile.
 */
class CarpoolInfoModal extends Component {

    /*
     * Purpose: call constructor of parent class and initializes the fields. 'state' contains 
     * the users that are in the carpool as well as a 'toggle' field which is a boolean 
     * that controls whether the modal is visible or not.
     */
    constructor(props) {
        super(props);
        this.toggle = this.toggle.bind(this);
        this.leaveCarpool = this.leaveCarpool.bind(this);
        this.users = app.database().ref().child('groupChats/'+this.props.carpoolID+"/users");

        this.state = {
            user:[],
            groupChatUsers:{},
            toggle: false
        };
    }

    /*
     * Purpose: acquires the users that are in the carpool and sets them to the state.
     */
    componentDidMount(){
        fetch('/api/account/profile/getAllUsers')
            .then(res => res.json())
            .then(json => this.setState({user: json}));
    }

    componentWillMount(){
        const previousUsers = this.state.groupChatUsers;

        this.users.on('child_added', snap =>{
            previousUsers[snap.key] = snap.val();
            this.setState({
                groupChatUsers: previousUsers
            });
        });
    }

    /*
     * Purpose: acquires the name of the user by iterating through the 'user' array in state
     * and matching the ID that is passed through to the function.
     */
    getUsername(_id) {

        for(let x in this.state.user) {

            if(this.state.user[x]._id === _id) {
                return this.state.user[x].firstName;
            }

        }

    }

    /*
     * Purpose: toggles whether the modal is visible or not by setting the 'toggle' field the 
     * opposite of the previous value. It called when the carpool name or the modal close button
     * is clicked.
     */
    toggle(event) {
        event.preventDefault();
        this.setState(prevState => ({
            toggle: !prevState.toggle
        }));
    }

    /*
     *Purpose: removes the user from the carpool and returns them to the homepage
     * It is called when the leave carpool button is clicked
     */
    leaveCarpool(){
        let tempUsers = {};
        for(let user in this.state.groupChatUsers){
            if(user !== getFromStorage('sessionKey').token){
                tempUsers[user] = this.state.groupChatUsers[user];
            }
        }

        this.setState({
            groupChatUsers: tempUsers
        }, function () {
            app.database().ref().child('groupChats/'+this.props.carpoolID)
                .update({users: this.state.groupChatUsers}).then(() => {
                return {};
            }).catch(error => {

                return {
                    errorCode: error.code,
                    errorMessage: error.message
                }

            });
        });

    }

    /*
     * Purpose: renders the component in the DOM. The visibility of the modal is dependant on the 'toggle' field.
     */
    render() {
        let users = [];

        for(let user in this.props.users) {
            users.push(
                <div className="row bordbot-1px-dash-grey" key={Math.random()}>
                    <div className="col-6 txt-left">{this.getUsername(user)}</div><div className="col-6 vertical-right"><Link to={"/ProfilePage/"+user}>View Profile</Link></div>
                </div>
            );
        }

        var modal = [];

        modal.push(
            // Modal
            <div key={Math.random()} className="modal" tabIndex="-1" role="dialog" id="carpoolInfoModal" style={this.state.toggle ? display : hide}>
                <div className="modal-dialog" role="document">
                    <div className="modal-content">
                        <div className="modal-header bg-aqua">
                            <h5 className="modal-title fw-bold">{this.props.carpoolName}</h5>
                            <button type="button" className="close" onClick={this.toggle} aria-label="Close">
                                <span aria-hidden="true">&times;</span>
                            </button>
                        </div>
                        <div className="modal-body">
                            <div className="row bordbot-1px-dash-grey">
                                <h6 className="fw-bold mx-auto">Carpool Members</h6>
                            </div>  
                            {users}
                        </div>
                        <a href="/HomePage" onClick={this.leaveCarpool}>
                            Leave carpool
                        </a>
                    </div>
                </div>
            </div>
        );

        return(
            <div className="col-8 txt-center">
                <button className="p-0 btn height-100p bg-trans txt-purple fw-bold brad-0 font-20px" onClick={this.toggle}>
                    {/* *** */}
                    {this.props.carpoolName}
                </button>
                {modal}
            </div>
        );
    }
}

export default CarpoolInfoModal;