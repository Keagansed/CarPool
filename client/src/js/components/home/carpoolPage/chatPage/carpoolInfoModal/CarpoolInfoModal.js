import React, { Component } from 'react';

const display = {
    display: 'block'
};
const hide = {
    display: 'none'
};

class CarpoolInfoModal extends Component{
    constructor(props){
        super(props);
        this.toggle = this.toggle.bind(this);

        this.state ={
            user:[],
            toggle: false
        };
    }

    componentDidMount(){
        fetch('/api/account/getAllUsers')
            .then(res => res.json())
            .then(json => this.setState({user: json}));
    }

    getUsername(_id)
    {
        for (var x in this.state.user)
        {
            if(this.state.user[x]._id === _id)
            {
                return this.state.user[x].firstName;
            }
        }

    }

    toggle(event) {
        this.setState(prevState => ({
            toggle: !prevState.toggle
        }));
    }

    render(){
        let users = [];

        for(let user in this.props.users)
        {
            users.push(
                <div className="row bordbot-1px-dash-grey" key={Math.random()}>
                    <div className="col-6">{this.getUsername(user)}</div><div className="col-6 vertical-right"><a href={"/ProfilePage/"+user}>View Profile</a></div>
                </div>
            );
        }

        var modal = [];
        modal.push(
            // Modal
            <div className="mx-auto" key={Math.random()}>
                <div key="0" className="modal" tabIndex="-1" role="dialog" id="carpoolInfoModal" style={this.state.toggle ? display : hide}>
                    <div className="modal-dialog" role="document">
                        <div className="modal-content">
                            <div className="modal-header bg-aqua">
                                <h5 className="modal-title fw-bold">{this.props.carpoolName}</h5>
                                <button type="button" className="close" onClick={this.toggle} aria-label="Close">
                                    <span aria-hidden="true">&times;</span>
                                </button>
                            </div>
                            <div className="modal-body">
                                <div className="row bordbot-1px-dash-grey">
                                    <h6 className="fw-bold mx-auto">Carpool Members</h6>
                                </div>
                                {users}
                            </div>
                        </div>
                    </div>
                </div>
            </div>
        );

        return (
            <div className="mx-auto">
                <button className="col-8 btn height-100p bg-trans txt-purple fw-bold brad-0 font-20px" onClick={this.toggle}>
                    {/* *** */}
                    {this.props.carpoolName}
                </button>
                {modal}
            </div>
        );
    }
}

export default CarpoolInfoModal;