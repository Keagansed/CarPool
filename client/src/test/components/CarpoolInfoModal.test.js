import React from 'react';
import CarpoolInfoModal from '../../js/components/carpool/CarpoolInfoModal';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

describe('CarpoolInfoModal Component', () => {

    it ('captures snapshot', () => {
        const renderedValue = renderer.create(<CarpoolInfoModal />).toJSON()
        expect(renderedValue).toMatchSnapshot();
    });

    it ('renders correctly', () => {
        let wrapper = shallow(<CarpoolInfoModal />)
        expect(wrapper.length).toEqual(1);
    });
});