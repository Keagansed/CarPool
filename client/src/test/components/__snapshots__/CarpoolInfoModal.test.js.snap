// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarpoolInfoModal Component captures snapshot 1`] = `
<div
  className="col-8 txt-center"
>
  <button
    className="p-0 btn height-100p bg-trans txt-purple fw-bold brad-0 font-20px"
    onClick={[Function]}
  />
  <div
    className="modal"
    id="carpoolInfoModal"
    role="dialog"
    style={
      Object {
        "display": "none",
      }
    }
    tabIndex="-1"
  >
    <div
      className="modal-dialog"
      role="document"
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header bg-aqua"
        >
          <h5
            className="modal-title fw-bold"
          />
          <button
            aria-label="Close"
            className="close"
            onClick={[Function]}
            type="button"
          >
            <span
              aria-hidden="true"
            >
              Ã—
            </span>
          </button>
        </div>
        <div
          className="modal-body"
        >
          <div
            className="row bordbot-1px-dash-grey"
          >
            <h6
              className="fw-bold mx-auto"
            >
              Carpool Members
            </h6>
          </div>
          <div
            className="row padtop-10px m-0"
          >
            <button
              className="btn btn-primary mx-auto width-15rem brad-2rem bg-red txt-purple fw-bold"
              id="btnLeaveCarpool"
              onClick={[Function]}
            >
              Leave Carpool
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="modal"
    id="carpoolInfoModal"
    role="dialog"
    style={
      Object {
        "display": "none",
      }
    }
    tabIndex="-1"
  >
    <div
      className="modal-dialog"
      role="document"
    >
      <div
        className="modal-content"
      >
        <div
          className="modal-header bg-aqua"
        >
          <h5
            className="modal-title fw-bold"
          />
        </div>
        <div
          className="modal-body"
        >
          <div
            className="row"
          >
            <h6
              className="fw-bold mx-auto"
            >
              Are you sure you want to leave this carpool?
            </h6>
          </div>
          <div
            className="row"
          >
            <button
              className="col-5 btn btn-primary mx-auto width-15rem brad-2rem mbottom-0 bg-aqua txt-purple fw-bold"
              id="btnYesCancel"
              onClick={[Function]}
              type="submit"
            >
              Yes
            </button>
            <button
              className="col-5 btn btn-primary mx-auto width-15rem brad-2rem mbottom-0 bg-aqua txt-purple fw-bold"
              id="btnNoCancel"
              onClick={[Function]}
              type="submit"
            >
              No
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
